### PR TITLE
UAVCAN Rangefinder Support

### PR DESCRIPTION
**Pull request by avionicsanonymous**
**Describe problem solved by this pull request**
Add support for rangefinder connected via UAVCAN.

**Describe your solution**
Added sensor bridge for "rangefinder" sensor, which accepts the included AvAnon rangefinder message. This custom message is added because the standard UAVCAN rangefinder message does not include the necessary data to fill out the PX4 distance_sensor topic, such as the min and max range of the connected sensor.

**Describe possible alternatives**
The standard UAVCAN rangefinder message could be used, but we would have to come up with some other way to convey the limits of the rangefinder. Since I'm not aware of any currently-existing devices that use that message, this should be just as good. If any exist in the future, we can add handlers for that message to this sensor bridge and solve the missing data at that time. 

**Test data / coverage**
Bench and flight tested using the AvAnon rangefinder UAVCAN interface. "distance_sensor" data was logged as expected.

### CODE REVIEW

**Comment by avionicsanonymous** [comment]
I see the Jenkins SIL tests failed, but can't quite glean why, and can't fathom how CAN driver additions could cause only a tailsitter SIL test to fail anyway?

**Comment by hamishwillee** [comment]
Cool! Is there any particular rangefinder this has been tested with, and if so, can we have a page for that in the user guide? http://docs.px4.io/master/en/sensor/rangefinders.html

**Comment by dagar** [comment]
I'm not sure about carrying custom DSDL for this. What about using `uavcan::equipment::range_sensor::Measurement` and instead we add parameters to set the min and max?

**Comment by avionicsanonymous** [comment]
> 
> 
> I'm not sure about carrying custom DSDL for this. What about using `uavcan::equipment::range_sensor::Measurement` and instead we add parameters to set the min and max?

I considered something like that, but it seemed more counter to the existing scheme where the user shouldn't have to sort that out themselves. Also, what if there are multiple rangefinders with different ranges? Then you'd not only need multiple sets of params, but those values become node-id dependent. One of the advantages of the CAN-connected rangefinder is that it enables multiple/redundant sensors, so limiting it to one would be a shame.

Thoughts?

**Comment by avionicsanonymous** [comment]
> 
> 
> Cool! Is there any particular rangefinder this has been tested with, and if so, can we have a page for that in the user guide? http://docs.px4.io/master/en/sensor/rangefinders.html

It was developed and tested around this unit, which acts as an interface between the autopilot any many common rangefinders: https://www.tindie.com/products/avionicsanonymous/uavcan-laser-altimeter-interface/

**Comment by dagar** [comment]
> I considered something like that, but it seemed more counter to the existing scheme where the user shouldn't have to sort that out themselves. Also, what if there are multiple rangefinders with different ranges? Then you'd not only need multiple sets of params, but those values become node-id dependent. One of the advantages of the CAN-connected rangefinder is that it enables multiple/redundant sensors, so limiting it to one would be a shame.
> 
> Thoughts?

I agree with everything you said, I'm just trying to weigh it against the potential harm of custom DSDL for something relatively common. The quick and dirty solution is to carry that metadata in the PX4 uavcan module, then if the need arises for more than one rangefinder over UAVCAN with different metadata we'd have to invent something. 

Then we make sure we get it right for UAVCANv1 (a large ongoing effort).

**Comment by avionicsanonymous** [comment]
> 
> 
> > I considered something like that, but it seemed more counter to the existing scheme where the user shouldn't have to sort that out themselves. Also, what if there are multiple rangefinders with different ranges? Then you'd not only need multiple sets of params, but those values become node-id dependent. One of the advantages of the CAN-connected rangefinder is that it enables multiple/redundant sensors, so limiting it to one would be a shame.
> > Thoughts?
> 
> I agree with everything you said, I'm just trying to weigh it against the potential harm of custom DSDL for something relatively common. The quick and dirty solution is to carry that metadata in the PX4 uavcan module, then if the need arises for more than one rangefinder over UAVCAN with different metadata we'd have to invent something.
> 
> Then we make sure we get it right for UAVCANv1 (a large ongoing effort).

Fair enough - and I suppose the most likely scenario if you have redundant rangefinders is that they're identical anyway. I'll make those changes.

**Comment by avionicsanonymous** [comment]
@dagar , where do you suggest putting the param definitions? Within the UAVCAN driver param set (which are currently mostly CAN server-centric), with the sensor params, or...?

**Comment by hamishwillee** [comment]
When this is in it will need documentation. At high level that means "enough for someone else to use it". 
At lower level that usually means a page [below here](https://docs.px4.io/master/en/sensor/rangefinders.html) that explains how to set up the rangefinder of the particular type.

You might also add info in the devguide if there are particular things that someone wanting to add uavcan support to their own rangefinder needs to know.

**Comment by avionicsanonymous** [comment]
@hamishwillee , Thanks for the suggestion. I have the documentation ready to go. I'll put in a PR for that once this is merged.

@dagar , I think this is good to go. Let me know if you have any other concerns!

File: src/drivers/uavcan/sensors/rangefinder.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	return 0;
+}
+
+void
+UavcanRangefinderBridge::range_sub_cb(const
+				      uavcan::ReceivedDataStructure<uavcan::equipment::range_sensor::Measurement> &msg)
+{
+	distance_sensor_s report{};
```
Comment (JacobCrabill) [inline src/drivers/uavcan/sensors/rangefinder.cpp:77]: Wouldn't it be neater to use PX4Rangefinder to handle the uORB publishing?

Comment (avionicsanonymous) [reply src/drivers/uavcan/sensors/rangefinder.cpp]: Well, I'm not sure if it's cleaner, but it's done!

**Comment by avionicsanonymous** [comment]
Rebased on current master and refactored to use PX4Rangefinder.

File: src/drivers/uavcan/sensors/rangefinder.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	if (rangefinder == nullptr) {
+		return;
+	}
+
+	if (!_inited) {
+
+		rangefinder->set_fov(msg.field_of_view);
+		rangefinder->set_min_distance(_range_min_m);
+		rangefinder->set_max_distance(_range_max_m);
+
```
[the remark this answers is not in the document]
Comment (avionicsanonymous) [reply src/drivers/uavcan/sensors/rangefinder.cpp:115]: Thanks! I think that slipped through the cracks when I refactored to use PX4Rangefinder. I added that to the latest push.

**Comment by avionicsanonymous** [comment]
@dagar do you want to see anything else or is this ready to merge?

**Comment by avionicsanonymous** [comment]
@dagar sorry to poke you again, but when can we merge this? I have a few of these devices in the wild with users who don't build from source but want to be able to use them!

**Comment by hamishwillee** [comment]
@avionicsanonymous Do you send max/min range in UAVCAN somehow? I ask because of this question: https://github.com/PX4/Firmware/pull/15551

**Comment by avionicsanonymous** [comment]
> 
> 
> @avionicsanonymous Do you send max/min range in UAVCAN somehow? I ask because of this question: #15551

Yes - this PR includes two new parameters, UAVCAN_RNG_MIN and UAVCAN_RANGE_MAX, which are to be set to match the rangefinder's min and max range. I believe I addressed this in the WIP documentation PR as well, but will go make sure of that.

Think we can get this merged soon? Evidently I'm not the only one who needs it!

**Comment by hamishwillee** [comment]
@avionicsanonymous Thanks

> Think we can get this merged soon? Evidently I'm not the only one who needs it!

That's up to @dagar or some other member of the core dev team. Dan, is there someone else who might review the changes to get this in?

**Comment by avionicsanonymous** [comment]
Updated to fix merge conflicts with master, which keeps running away from this PR. A couple sim tests are failing but I doubt that has anything to do with the relevant changes here.

@dagar , can we please merge this soon so I don't have to keep fixing it to stay up to date?

**Comment by avionicsanonymous** [comment]
@dagar , what can I do to help get this merged? Thanks!

**Comment by dagar** [comment]
> @dagar , what can I do to help get this merged? Thanks!

Sorry for the delay, I missed that the issue with using custom DSDL had been resolved. Let's get this in and make sure that the proper metadata is in place as we get closer to UAVCANv1.

**Pull request merged**